lib/live_view_native/extensions/modifiers.ex

```diff
@@ -6,41 +6,179 @@ defmodule LiveViewNative.Extensions.Modifiers do
+          # the same arguments as the `params` function, calling that function before applying
+          # a modifier to the stack.
+
+          def unquote(:"#{modifier_key}")({:_apply_mod, {mod_args, mod_builder, opts}}) do
+            mod_builder = mod_builder || struct(unquote(modifiers_struct), %{})
+            raw_params = Keyword.get(opts, :raw_params)
+
+            params =
+              if raw_params,
+                do: raw_params,
+                else: apply(unquote(modifier_module), :params, mod_args)
+
             modifier = apply(unquote(modifier_module), :modifier, [params])
 
-            if modifiers do
-              Map.put(
-                ctx,
-                :modifiers,
-                LiveViewNativePlatform.ModifiersStack.append(modifiers, modifier)
-              )
-            else
-              ctx
+            ModifiersStack.append(mod_builder, modifier)
+          end
+
+          if Keyword.has_key?(modifier_module.__info__(:functions), :params) do
+            for n <- 0..max_constructor_arity do
+              # `params/n` for modifier `f` produces:
+              # `f/(n + 1)` - taking a context struct
+              # `f/(n + 1)` - taking a modifier builder
+              # `f/n` - taking no arguments
+              if Kernel.function_exported?(modifier_module, :params, n) do
+                args = Macro.generate_arguments(n, nil)
+
+                def unquote(:"#{modifier_key}")(%LiveViewNativePlatform.Context{modifiers: mod}, unquote_splicing(args)),
+                  do: unquote(:"#{modifier_key}")({:_apply_mod, {[unquote_splicing(args)], mod, []}})
+
+                def unquote(:"#{modifier_key}")(%unquote(modifiers_struct){} = mod, unquote_splicing(args)),
+                  do: unquote(:"#{modifier_key}")({:_apply_mod, {[unquote_splicing(args)], mod, []}})
+
+                def unquote(:"#{modifier_key}")(unquote_splicing(args)),
+                  do: unquote(:"#{modifier_key}")({:_apply_mod, {[unquote_splicing(args)], nil, []}})
+              end
+            end
+          else
+            # If no `params` callback is exported by the modifier module, only include one and two arity
+            # versions that optionally take a platform or modifier builder struct and always an option list,
+            # map or modifier struct. Zero arity modifier functions are also included for modifiers without
+            # schema fields.
+            def unquote(:"#{modifier_key}")(
+                  %LiveViewNativePlatform.Context{modifiers: modifiers},
+                  params
+                ) do
+              mod = struct(unquote(modifiers_struct), Map.from_struct(modifiers))
+
+              unquote(:"#{modifier_key}")(mod, params)
+            end
+
+            def unquote(:"#{modifier_key}")(%unquote(modifiers_struct){} = mod, params) do
+              modifier = apply(unquote(modifier_module), :modifier, [params])
```

Comment:
I'd have written this as:

`modifier = unquote(modifier_module).modifier(params)` to ensure that a compile-time failure happens if `modifier_module` doesn't have `modifier/1` implemented.

Reply:
@polvalente That's fair. Modifier modules will always define `modifier/1` since its pulled in by the `LiveViewNativePlatform.Modifier` macro they all use ([link](https://github.com/liveview-native/live_view_native_platform/blob/main/lib/live_view_native_platform/modifier.ex#L42) and [example of a modifier module](https://github.com/liveview-native/liveview-client-swiftui/blob/main/lib/live_view_native_swift_ui/modifiers/drawing_and_graphics/foreground_color.ex)) but additional compile-time safety never hurts. I'll improve this in a future PR, or feel free to submit one if you'd like 🙂